src/hotspot/share/opto/escape.cpp

```diff
@@ -2880,6 +2881,21 @@ bool ConnectionGraph::not_global_escape(Node *n) {
   return true;
 }
 
+// Return true if locked object does not escape globally
+// and locked code region (identified by BoxLockNode) is balanced:
+// all compiled code paths have corresponding Lock/Unlock pairs.
+bool ConnectionGraph::can_eliminate_lock(AbstractLockNode* alock) {
+  BoxLockNode* box = alock->box_node()->as_BoxLock();
+  if (!box->is_unbalanced() && not_global_escape(alock->obj_node())) {
+    if (EliminateNestedLocks) {
+      // We can mark whole locking region as Local only when only
+      // one object is used for locking.
+      box->set_local();
```

Comment:
Could the box be `nested` at this point? If so, the assert inside `set_local` could trigger, right?

Reply:
> Could the box be nested at this point? If so, the assert inside set_local could trigger, right?

No, it can't be nested.
Check for nesting and its elimination is done in one place from which it does not escape [macro.cpp#L2047](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/macro.cpp#L2047)
(Nested state is set only when `is_nested_lock_region()` returns true in the check).
And it will be overwritten immediately by `box_node->set_eliminated()`.

Reply:
Also BoxLock can't be Local when we check for nesting because of `(!alock->is_non_esc_obj())` check above.